<center>
    <img src="https://media-exp1.licdn.com/dms/image/C4E1BAQErMN84hgIX3Q/company-background_10000/0/1622546957980?e=2147483647&v=beta&t=l7tU8QXWhNOtm4NZ3ZLXRSw42o0U-QZu-3JbvBr94Fc" width="1000" alt="cognitiveclass.ai logo"  />
</center>

<h1 style="text-align:center; font-size: 5em;">Working with APIs</h1>

## Working with APIS
<div class="alert alert-block alert-danger">
    <b>Number one key when working with APIs is to read the documentation. APIs will have different parameters and structure, depending on how it was created.</b>
</div>

<div class="alert alert-block alert-warning">
    <b>Task:</b> create a <code>config.py</code> file to hold your API keys.
    <li> Reason: API Keys can cost money per number of calls. If we expose our keys on the code it can be used by unknown parties. 
</div>

`python -m pip install requests`

In [1]:
# Dependencies
import requests
from pprint import pprint
import json
import pandas as pd

In [1]:
import sys
import os

# Add the parent directory to sys.path
parent_dir = os.path.abspath(os.path.join(os.getcwd(), '..'))
sys.path.insert(0, parent_dir)

# Now you can import config.py
#from config import movies_api_key

# Use the config values
#print(movies_api_key)

<a name="anchorQueryBasics" style="position:absolute;"></a>
<hr style="border:2px solid">

# Query Basics
<hr style="border-top:1px dashed">

[Requests Documentation](https://docs.python-requests.org/en/latest/)

[Movie Info API](http://www.omdbapi.com/)

In [3]:
##Validate your API key:
api_key = movies_api_key

In [ ]:
##Movie Search
##Setting the parameters for the search


call to send to the API: http://www.omdbapi.com/?t=Casino Royale&apikey=c3ba08dc


In [ ]:

##If you get "<Response [200]>" then everything is fine, otherwise review your code, specially your API key.

<Response [200]>

In [ ]:
##Print the queried url. NOTE: Make sure to comment out as it will also print the API key


http://www.omdbapi.com/?t=Casino%20Royale&apikey=c3ba08dc


#### URL validation:

You can also validate your response by copying the URL from above and past it to a web browser.

__Is it similar to any the Python datatypes that you know?__ 


In [ ]:
# Check status code. A 200 means everything is ok with the syntax of the query.


200

**2XX Codes**<br>
<div class="alert alert-block alert-warning">
   These codes convey that the server has received the client’s request and processed it successfully. Here are the most common examples:
    <br>
</div>

**200 OK**: The request was successful.<br>
**201 Created**: The resource has been created on the server. This response is typically returned for POST requests.<br>
**202 Accepted**: The request has been received but is still being processed.<br>
**204 No Content**: The request has been successfully processed but no content will be returned.

'utf-8'

## Data analysis

In [ ]:
##Looking at the different functions and attribute of "response"


Help on Response in module requests.models object:

class Response(builtins.object)
 |  The :class:`Response <Response>` object, which contains a
 |  server's response to an HTTP request.
 |  
 |  Methods defined here:
 |  
 |  __bool__(self)
 |      Returns True if :attr:`status_code` is less than 400.
 |      
 |      This attribute checks if the status code of the response is between
 |      400 and 600 to see if there was a client error or a server error. If
 |      the status code, is between 200 and 400, this will return True. This
 |      is **not** a check to see if the response code is ``200 OK``.
 |  
 |  __enter__(self)
 |  
 |  __exit__(self, *args)
 |  
 |  __getstate__(self)
 |  
 |  __init__(self)
 |      Initialize self.  See help(type(self)) for accurate signature.
 |  
 |  __iter__(self)
 |      Allows you to use a response as an iterator.
 |  
 |  __nonzero__(self)
 |      Returns True if :attr:`status_code` is less than 400.
 |      
 |      This attribute checks if

In [ ]:
##Accessing the content of the response:


b'{"Title":"Casino Royale","Year":"2006","Rated":"PG-13","Released":"17 Nov 2006","Runtime":"144 min","Genre":"Action, Adventure, Thriller","Director":"Martin Campbell","Writer":"Neal Purvis, Robert Wade, Paul Haggis","Actors":"Daniel Craig, Eva Green, Judi Dench","Plot":"After earning a licence to kill, secret agent James Bond sets out on his first mission as 007. Bond must defeat a private banker funding terrorists in a high-stakes game of poker at Casino Royale, in Montenegro.","Language":"English, Serbian, German, Italian, French","Country":"United States, United Kingdom, Czech Republic, Germany, Bahamas, Italy","Awards":"Won 1 BAFTA Award28 wins & 44 nominations total","Poster":"https://m.media-amazon.com/images/M/MV5BMWQ1ZDM4NDktMWY0NC00MjcxLWJlMDMtNmE2MGVhYzRjMWQ0XkEyXkFqcGc@._V1_SX300.jpg","Ratings":[{"Source":"Internet Movie Database","Value":"8.0/10"},{"Source":"Rotten Tomatoes","Value":"94%"},{"Source":"Metacritic","Value":"80/100"}],"Metascore":"80","imdbRating":"8.0","imdb

In [ ]:
# Server's information for this request


{'Date': 'Mon, 20 Oct 2025 10:53:03 GMT', 'Content-Type': 'application/json; charset=utf-8', 'Transfer-Encoding': 'chunked', 'Connection': 'keep-alive', 'Cache-Control': 'public, max-age=3600', 'Expires': 'Mon, 20 Oct 2025 11:53:03 GMT', 'Last-Modified': 'Mon, 20 Oct 2025 10:53:03 GMT', 'Vary': '*', 'Server': 'cloudflare', 'X-AspNet-Version': '4.0.30319', 'Access-Control-Allow-Origin': '*', 'cf-cache-status': 'DYNAMIC', 'Content-Encoding': 'gzip', 'CF-RAY': '9917f81dafa30765-MAN'}

In [ ]:
# Convert the response to JSON format


<class 'dict'>


In [ ]:
##Pprint helps us to visualise better a JSON/dictionary data


{'Actors': 'Daniel Craig, Eva Green, Judi Dench',
 'Awards': 'Won 1 BAFTA Award28 wins & 44 nominations total',
 'BoxOffice': '$167,445,960',
 'Country': 'United States, United Kingdom, Czech Republic, Germany, Bahamas, '
            'Italy',
 'DVD': 'N/A',
 'Director': 'Martin Campbell',
 'Genre': 'Action, Adventure, Thriller',
 'Language': 'English, Serbian, German, Italian, French',
 'Metascore': '80',
 'Plot': 'After earning a licence to kill, secret agent James Bond sets out on '
         'his first mission as 007. Bond must defeat a private banker funding '
         'terrorists in a high-stakes game of poker at Casino Royale, in '
         'Montenegro.',
 'Poster': 'https://m.media-amazon.com/images/M/MV5BMWQ1ZDM4NDktMWY0NC00MjcxLWJlMDMtNmE2MGVhYzRjMWQ0XkEyXkFqcGc@._V1_SX300.jpg',
 'Production': 'N/A',
 'Rated': 'PG-13',
 'Ratings': [{'Source': 'Internet Movie Database', 'Value': '8.0/10'},
             {'Source': 'Rotten Tomatoes', 'Value': '94%'},
             {'Source': 'Metac

In [ ]:
# Retriveve all the Keys


dict_keys(['Title', 'Year', 'Rated', 'Released', 'Runtime', 'Genre', 'Director', 'Writer', 'Actors', 'Plot', 'Language', 'Country', 'Awards', 'Poster', 'Ratings', 'Metascore', 'imdbRating', 'imdbVotes', 'imdbID', 'Type', 'DVD', 'BoxOffice', 'Production', 'Website', 'Response'])

In [ ]:
# Retriveve all Values


dict_values(['Casino Royale', '2006', 'PG-13', '17 Nov 2006', '144 min', 'Action, Adventure, Thriller', 'Martin Campbell', 'Neal Purvis, Robert Wade, Paul Haggis', 'Daniel Craig, Eva Green, Judi Dench', 'After earning a licence to kill, secret agent James Bond sets out on his first mission as 007. Bond must defeat a private banker funding terrorists in a high-stakes game of poker at Casino Royale, in Montenegro.', 'English, Serbian, German, Italian, French', 'United States, United Kingdom, Czech Republic, Germany, Bahamas, Italy', 'Won 1 BAFTA Award28 wins & 44 nominations total', 'https://m.media-amazon.com/images/M/MV5BMWQ1ZDM4NDktMWY0NC00MjcxLWJlMDMtNmE2MGVhYzRjMWQ0XkEyXkFqcGc@._V1_SX300.jpg', [{'Source': 'Internet Movie Database', 'Value': '8.0/10'}, {'Source': 'Rotten Tomatoes', 'Value': '94%'}, {'Source': 'Metacritic', 'Value': '80/100'}], '80', '8.0', '726,077', 'tt0381061', 'movie', 'N/A', '$167,445,960', 'N/A', 'N/A', 'True'])

### Navigating throught the "response"

Because the data is now a dictionary, it might the helpful to review the session where we saw this subject. 

In [ ]:
# Get individual Responses from the JSON file


Movie director Martin Campbell
Amount made at the box office $167,445,960


In [ ]:
# From multiple values
## Lets analyse the "rating" of the movie
####First level:


<class 'list'>


[{'Source': 'Internet Movie Database', 'Value': '8.0/10'},
 {'Source': 'Rotten Tomatoes', 'Value': '94%'},
 {'Source': 'Metacritic', 'Value': '80/100'}]

In [ ]:
# From multiple values

##Accessing the first element of the list:


<class 'dict'>
{'Source': 'Internet Movie Database', 'Value': '8.0/10'}


In [ ]:
##Accessing the key, value pair:

##Raiting source

##Raiting:


dict_keys(['Source', 'Value'])
Internet Movie Database
8.0/10


__Exercise:__ Print all the raitings available in "response"

In [ ]:
##Your code:


Internet Movie Database: 8.0/10
Rotten Tomatoes: 94%
Metacritic: 80/100


__Exercise:__ Create a function that, given an user input, returns the following data (e.g. movie UP): 

In [ ]:
## Your code here: 



call to send to the API: http://www.omdbapi.com/?t=UP&apikey=c3ba08dc
Movie Title: Up
Director: Pete Docter, Bob Peterson
Box Office: $293,004,164
Raitings:
Internet Movie Database: 8.3/10
Rotten Tomatoes: 98%
Metacritic: 88/100


<a name="anchorBadQuery" style="position:absolute;"></a>
<hr style="border:2px solid">

# Bad Query
<hr style="border-top:1px dashed">

In [22]:
# Look for all movies
url2 = 'http://www.omdbapi.com/?'
bad_key = '1234'
api = 'apikey='
year = "&y="+ '2019'


In [23]:
## Bad API KEY
#query= url2 + api+ bad_key + year
# Not Formatted properly
query= url2 + api_key + year

In [24]:
print(query)

http://www.omdbapi.com/?c3ba08dc&y=2019


In [25]:
# Performing a GET request 
response2 = requests.get(query)
print(response2)
# Response 400 means it is a bad request

<Response [401]>


In [26]:
# Convert the response to JSON format
action_data = response2.json()
pprint(action_data)

{'Error': 'No API key provided.', 'Response': 'False'}


In [ ]:
##Movie Search
##Setting the parameters for the search
# Look for all movies


'http://www.omdbapi.com/?c3ba08dc&y=2019'

The documentation clearly says that one of the parameters i or t are required for a query.

**4XX Codes**<br>
<div class="alert alert-block alert-warning">
   These codes convey that the client’s request has an error. It’s possible that the request is written incorrectly, or the resource which the client is requesting doesn't exist.
    <br>
</div>

**400 Bad Request**: There is something wrong in the client’s request.<br>
**401 Unauthorized**: The client is not authorized to make this request.<br>
**403 Forbidden**: The request is correct but can’t be processed. Likely, the problem is that the client does not have required permissions.<br>
**404 Not Found**: The resource being requested doesn't exist.

<a name="anchorPag" style="position:absolute;"></a>
<hr style="border:2px solid">

# Pagination
<hr style="border-top:1px dashed">

<div class="alert alert-block alert-warning">
    <b>What does it do?:</b> Pagination limits the number of results to help keep network traffic in check.
    <br>We should read the API's documentation to understand how many values we have per page. 
</div>

Source: https://datahelpdesk.worldbank.org/knowledgebase/articles/898590-country-api-queries

In [29]:
url3 = "http://api.worldbank.org/v2/"
api_format = "json"
query2= f"{url3}countries?format={api_format}"
query2


'http://api.worldbank.org/v2/countries?format=json'

In [30]:
###Validating the reply:
requests.get(query2)

<Response [200]>

In [ ]:
# Get country information in JSON format


<class 'list'> 2


[{'page': 1, 'pages': 6, 'per_page': '50', 'total': 296},
 [{'id': 'ABW',
   'iso2Code': 'AW',
   'name': 'Aruba',
   'region': {'id': 'LCN',
    'iso2code': 'ZJ',
    'value': 'Latin America & Caribbean '},
   'adminregion': {'id': '', 'iso2code': '', 'value': ''},
   'incomeLevel': {'id': 'HIC', 'iso2code': 'XD', 'value': 'High income'},
   'lendingType': {'id': 'LNX', 'iso2code': 'XX', 'value': 'Not classified'},
   'capitalCity': 'Oranjestad',
   'longitude': '-70.0167',
   'latitude': '12.5167'},
  {'id': 'AFE',
   'iso2Code': 'ZH',
   'name': 'Africa Eastern and Southern',
   'region': {'id': 'NA', 'iso2code': 'NA', 'value': 'Aggregates'},
   'adminregion': {'id': '', 'iso2code': '', 'value': ''},
   'incomeLevel': {'id': 'NA', 'iso2code': 'NA', 'value': 'Aggregates'},
   'lendingType': {'id': '', 'iso2code': '', 'value': 'Aggregates'},
   'capitalCity': '',
   'longitude': '',
   'latitude': ''},
  {'id': 'AFG',
   'iso2Code': 'AF',
   'name': 'Afghanistan',
   'region': {'id': 

In [ ]:
# First element is general information, second is countries themselves


{'page': 1, 'pages': 6, 'per_page': '50', 'total': 296}


[{'id': 'ABW',
  'iso2Code': 'AW',
  'name': 'Aruba',
  'region': {'id': 'LCN',
   'iso2code': 'ZJ',
   'value': 'Latin America & Caribbean '},
  'adminregion': {'id': '', 'iso2code': '', 'value': ''},
  'incomeLevel': {'id': 'HIC', 'iso2code': 'XD', 'value': 'High income'},
  'lendingType': {'id': 'LNX', 'iso2code': 'XX', 'value': 'Not classified'},
  'capitalCity': 'Oranjestad',
  'longitude': '-70.0167',
  'latitude': '12.5167'},
 {'id': 'AFE',
  'iso2Code': 'ZH',
  'name': 'Africa Eastern and Southern',
  'region': {'id': 'NA', 'iso2code': 'NA', 'value': 'Aggregates'},
  'adminregion': {'id': '', 'iso2code': '', 'value': ''},
  'incomeLevel': {'id': 'NA', 'iso2code': 'NA', 'value': 'Aggregates'},
  'lendingType': {'id': '', 'iso2code': '', 'value': 'Aggregates'},
  'capitalCity': '',
  'longitude': '',
  'latitude': ''},
 {'id': 'AFG',
  'iso2Code': 'AF',
  'name': 'Afghanistan',
  'region': {'id': 'MEA',
   'iso2code': 'ZQ',
   'value': 'Middle East, North Africa, Afghanistan & Pa

In [ ]:
# Check number of responses


Number of responses 50


In [ ]:
##We can see that there are more elements to extract:
## 'page': 1, 'pages': 6, 'per_page': '50


{'page': 1, 'pages': 6, 'per_page': '50', 'total': 296}

In [ ]:
##Requesting the second page:


'http://api.worldbank.org/v2/countries?&page=1&format=json'

In [ ]:
##Testing the amount of pages that are in the data source
##Lets see if there is a page = 20



<Response [200]>

In [ ]:
##But accessing the actual data we can see that has no countries info.


{'page': 20, 'pages': 6, 'per_page': '50', 'total': 296}


[]

[{'page': 20, 'pages': 6, 'per_page': '50', 'total': 296}, []]

__Exercise:__ Taking in consideration the previous points, develop a code that will collect all the coutries information from the data source. Do not hardcode the solution, it has to be dynamic.

In [ ]:
##Your code here:



http://api.worldbank.org/v2/countries?&page=1&format=json
http://api.worldbank.org/v2/countries?&page=2&format=json
http://api.worldbank.org/v2/countries?&page=3&format=json
http://api.worldbank.org/v2/countries?&page=4&format=json
http://api.worldbank.org/v2/countries?&page=5&format=json
http://api.worldbank.org/v2/countries?&page=6&format=json
http://api.worldbank.org/v2/countries?&page=7&format=json
Items extracted:  296
Data sample:
[{'adminregion': {'id': '', 'iso2code': '', 'value': ''},
  'capitalCity': 'Oranjestad',
  'id': 'ABW',
  'incomeLevel': {'id': 'HIC', 'iso2code': 'XD', 'value': 'High income'},
  'iso2Code': 'AW',
  'latitude': '12.5167',
  'lendingType': {'id': 'LNX', 'iso2code': 'XX', 'value': 'Not classified'},
  'longitude': '-70.0167',
  'name': 'Aruba',
  'region': {'id': 'LCN',
             'iso2code': 'ZJ',
             'value': 'Latin America & Caribbean '}},
 {'adminregion': {'id': '', 'iso2code': '', 'value': ''},
  'capitalCity': '',
  'id': 'AFE',
  'incom

__Exercise:__ From the information extracted, create the following dataframe: 

In [ ]:
##Your code here:


country           income_lvl
0                        Aruba          High income
1  Africa Eastern and Southern           Aggregates
2                  Afghanistan           Low income
3                       Africa           Aggregates
4   Africa Western and Central           Aggregates
5                       Angola  Lower middle income
6                      Albania  Upper middle income
7                      Andorra          High income
8                   Arab World           Aggregates
9         United Arab Emirates          High income